### PR TITLE
feat: Detray plugin

### PR DESCRIPTION
**Pull request by EleniXoch**
Detray Plugin feature: 

  -Compile detray in acts as thirdparty(cmake flag ACTS_SETUP_DETRAY)
  -New detray_plugin flag introduction for integration (ACTS_BUILD_PLUGIN_DETRAY )
  -Continuing conversion of detray geometry to avoid json: from acts detector structure to detray payloads to converted detray detector

Status:
  -Portals, volumes, surfaces completed (replacing odd-detray_geometry_detray.json)
  -Surface grids and material conversion in progress
  -host_mr issue to be resolved(temporary solution)

Conversion implementation for populating payloads in Cpp under acts/Plugins/Detray
Pybind implementaion for calling conversion from python.

Compile with:

```
cmake -S acts -B buildD/acts -DCMAKE_INSTALL_PREFIX=installD/acts -DACTS_BUILD_EVERYTHING=ON -DACTS_BUILD_ODD=ON -DACTS_BUILD_EXAMPLES_PYTHON_BINDINGS=ON -DACTS_BUILD_PLUGIN_ACTSVG=ON -DACTS_BUILD_PLUGIN_DETRAY=ON
cmake --build buildD/acts
```

Run in detector_creation.py: run acts.examples.DetrayConverter(geoContext, detector,“odd-detray”)

### CODE REVIEW

File: thirdparty/OpenDataDetector
Comment (asalzburger) [inline thirdparty/OpenDataDetector:1]: Can you revert this to the version in main please ?

File: Examples/Python/src/Detray.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:3]:
```suggestion
// Copyright (C) 2024 CERN for the benefit of the Acts project
```

File: Examples/Python/src/Detray.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Definitions/Algebra.hpp"
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:9]:
```suggestion
```

File: Examples/Python/src/Detray.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#include "Acts/Detector/Detector.hpp"
+#include "Acts/Detector/ProtoDetector.hpp"
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:11]:
```suggestion
```

File: Examples/Python/src/Detray.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+#include "Acts/Plugins/Python/Utilities.hpp"
+#include "Acts/Utilities/Logger.hpp"
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:13]:
```suggestion
```

File: Examples/Python/src/Detray.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/operators.h>
+
+namespace Acts {
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:38]: All these lines are not needed

File: Examples/Python/src/Detray.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+#include "detray/io/frontend/implementation/json_readers.hpp"
+#include "detray/io/frontend/utils/detector_components_reader.hpp"
+#include "detray/utils/consistency_checker.hpp"
+
+#include "Detray.hpp"
+
+#include <vecmem/memory/memory_resource.hpp>
+#include <vecmem/memory/host_memory_resource.hpp>
+
+#include <fstream>
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:26]:
```suggestion
```

File: Examples/Python/src/Detray.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+#include <initializer_list>
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:27]:
```suggestion
```

File: Examples/Python/src/Detray.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+#include <memory>
+#include <string>
+#include <tuple>
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:30]:
```suggestion
```

File: Examples/Python/src/Detray.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+#include <vector>
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:31]:
```suggestion
```

File: Examples/Python/src/Detray.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+#include <nlohmann/json.hpp>
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:33]:
```suggestion
```

File: Examples/Python/src/Detray.cpp [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                    [](const Acts::GeometryContext& gctx,
+                    const Acts::Experimental::Detector& acts_detector,
+                    const std::string& name) -> auto {//detector_t
+                        
+                        // Create a host memory resource
+                        vecmem::host_memory_resource host_mr;
+                        // Convert Acts detector to detray detector using the detray_tree_converter function
+                        auto d_detray = detray_tree_converter(acts_detector, gctx, host_mr);   
+                        
+                        return true;//TO DO:: return d_detray; after host_mr is fixed
```
Comment (asalzburger) [inline Examples/Python/src/Detray.cpp:83]: What we discussed here is that on the ACTS (at least examples side), we declare a `detray` object always as a tuple of detector and memory source, so you could change here to return both, that should fix the problem?

File: Plugins/CMakeLists.txt [hunk 12]
```diff
@@ -10,6 +10,7 @@ add_component_if(Legacy PluginLegacy ACTS_BUILD_PLUGIN_LEGACY)
 add_component_if(Onnx PluginOnnx ACTS_BUILD_PLUGIN_ONNX)
 add_component_if(Sycl PluginSycl ACTS_BUILD_PLUGIN_SYCL)
 add_component_if(ExaTrkX PluginExaTrkX ACTS_BUILD_PLUGIN_EXATRKX)
+add_component_if(Detray PluginDetray ACTS_BUILD_PLUGIN_DETRAY)
```
Comment (asalzburger) [inline Plugins/CMakeLists.txt:13]: We will have to align this with Frederik's work.

File: Plugins/Detray/include/Detray.hpp [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+    const std::vector<const Acts::Experimental::DetectorVolume*>& volumes) {
+  auto candidate = std::find(volumes.begin(), volumes.end(), volume);
+  if (candidate != volumes.end()) {
+    return std::distance(volumes.begin(), candidate);
+  }
+  return -1;
+}
+}  // namespace
+
+namespace detray{
```
Comment (asalzburger) [inline Plugins/Detray/include/Detray.hpp:107]: Shall we call this `namespace DetrayConverter` ?

File: Plugins/Detray/include/Detray.hpp [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+        typename detector_t::name_map names{};
+        out_json["data"] = detray::io::geometry_writer::convert(det, names);
+        outputFile << out_json << std::endl;
+        return;
+    }
+    
+    /// @return the transform_payload(translation, rotation) of each surface/volume
+    /// @param Transform3 acts object, Transform3JsonConverter::Options
+    /// @brief convert the acts transform to detray transform payload
+    static io::transform_payload detray_converter_transf(
```
Comment (asalzburger) [inline Plugins/Detray/include/Detray.hpp:126]: For code in Acts we use `CamelCase` nomenclature, i.e. this method would be called like `detrayConverterTransform` - however, if we called the namespace here `DetrayConverter`, then converting a transform would be:

`DetrayConverter::convertTransform()`

which is pretty explicit what it does.

File: Plugins/Detray/include/Detray.hpp [hunk 15]
```diff
@@ -0,0 +1,552 @@
...
+        } else {
+            p_acts.rot = {};
+        }
+
+        return p_acts;
+    }
+
+    /// @return the mask_payload of the surface @param bounds, @param portal
+    /// @brief convert the acts bounds to detray mask payload
+    static io::mask_payload detray_converter_mask(
```
Comment (asalzburger) [inline Plugins/Detray/include/Detray.hpp:157]: Same here, this would be `DetrayConverter::convertMask`

File: Plugins/Detray/include/Detray.hpp [hunk 16]
```diff
@@ -0,0 +1,552 @@
...
+        const Acts::SurfaceBounds& bounds, bool portal){
+        //Acts::SurfaceBoundsJsonConverter::toJsonDetray
+
+        detray::io::mask_payload mask_pd;
+        auto [shape, boundaries] = DetrayJsonHelper::maskFromBounds(bounds, portal);
```
Comment (asalzburger) [inline Plugins/Detray/include/Detray.hpp:162]: Probably we should shift the  `DetrayJsonHelper::maskFromBound` code to here and call it from the Json helper.

File: Plugins/Detray/include/Detray.hpp [hunk 17]
```diff
@@ -0,0 +1,552 @@
...
+        //TO DO: use inline single_link_payload convert(const std::size_t idx) 
+        detray::io::single_link_payload lnk;
+        mask_pd.volume_link = lnk;
+
+        return mask_pd;
+    }
+
+    /// @return the surface_payload for portals and volumes by @param Surface acts object
+    /// @brief convert the acts surface to detray surface payload and populate the payload
+    static io::surface_payload detray_converter_surf(
```
Comment (asalzburger) [inline Plugins/Detray/include/Detray.hpp:176]: Same naming here.

File: Plugins/Detray/src/Detray.cpp [hunk 18]
```diff
@@ -0,0 +1 @@
+
```
Comment (asalzburger) [inline Plugins/Detray/src/Detray.cpp:1]: Is this an empty file ?

**Review by asalzburger — CHANGES_REQUESTED** [review]
Great to see this coming in, here's a first brush over - I this we should try to split the coversion code into several logical files, in order to not make them too long.

File: Examples/Python/src/Detray.cpp (hunk 11 above)
Comment (EleniXoch) [reply Examples/Python/src/Detray.cpp:83]: Yes i will try it like that thanks :)

File: Plugins/Detray/include/Detray.hpp [hunk 19]
```diff
@@ -0,0 +1,552 @@
...
+
+namespace detray{
+
+    /// These functions perform the conversion of the ACTS detector to the detray detector. 
+    /// The payloads are created and populated by traversing the initial ACTS detector object.
+
+    /// detray geometry writer function, debug purposes 
+    void detray_detector_print(const detector_t& det){
+
+        std::ofstream outputFile("data_try.json");
```
Comment (niermann999) [inline Plugins/Detray/include/Detray.hpp:115]: This should call the detray detector writer, if possible:
```
#include "detray/io/frontend/detector_writer.hpp"
```
```
    auto writer_cfg = detray::io::detector_writer_config{}
                          .format(detray::io::format::json)
                          .replace_files(true);

    detray::io::write_detector(det, names, writer_cfg);
```

File: Plugins/Detray/include/Detray.hpp [hunk 20]
```diff
@@ -0,0 +1,552 @@
...
+#include "detray/io/frontend/implementation/json_readers.hpp"
+#include "detray/io/frontend/utils/detector_components_reader.hpp"
+
+#include "detray/io/common/geometry_reader.hpp"
+#include "detray/io/common/geometry_writer.hpp"
+
+#include "detray/io/json/json_reader.hpp"
+#include "detray/io/json/json_writer.hpp"
+#include "detray/io/json/json_io.hpp"
+#include "detray/io/frontend/writer_interface.hpp"
```
Comment (niermann999) [inline Plugins/Detray/include/Detray.hpp:46]: Are these still needed?

**Pull request closed without merging**